Train Step Compile [1/3] basic fwd/bwd only

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98461
* __->__ #100605

Adds infra to wrap compilers in TrainStepCompiler and logic to dynamo
to behave specially (e.g. allowing .backward() to not graph-break)
when using such compilers.

Integrates some state with TracingContext, useful for both asserting
only one call to .backard() as well as groundwork for optimizer support
in following PR.

Checks various assumptions about the user trainstep function:
 - only one call to .backward() - double backwards is uncommon, and
   initial users of train step compile are expected to be satisfied
   without it.  Not sure if it would be feasible to support.
 - only supported backends are allowed
 - fullgraph (nopython) is required when train_step_compiling

Separated PRs for easier review, but this is of minimal utility
on its own without optimizer support.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

wconstab committed:
Train Step Compile [1/3] basic fwd/bwd only

Adds infra to wrap compilers in TrainStepCompiler and logic to dynamo
to behave specially (e.g. allowing .backward() to not graph-break)
when using such compilers.

Integrates some state with TracingContext, useful for both asserting
only one call to .backard() as well as groundwork for optimizer support
in following PR.

Checks various assumptions about the user trainstep function:
 - only one call to .backward() - double backwards is uncommon, and
   initial users of train step compile are expected to be satisfied
   without it.  Not sure if it would be feasible to support.
 - only supported backends are allowed
 - fullgraph (nopython) is required when train_step_compiling

Separated PRs for easier review, but this is of minimal utility
on its own without optimizer support.

[ghstack-poisoned]

wconstab committed:
Update on "Train Step Compile [1/3] basic fwd/bwd only"


Adds infra to wrap compilers in TrainStepCompiler and logic to dynamo
to behave specially (e.g. allowing .backward() to not graph-break)
when using such compilers.

Integrates some state with TracingContext, useful for both asserting
only one call to .backard() as well as groundwork for optimizer support
in following PR.

Checks various assumptions about the user trainstep function:
 - only one call to .backward() - double backwards is uncommon, and
   initial users of train step compile are expected to be satisfied
   without it.  Not sure if it would be feasible to support.
 - only supported backends are allowed
 - fullgraph (nopython) is required when train_step_compiling

Separated PRs for easier review, but this is of minimal utility
on its own without optimizer support.

[ghstack-poisoned]

Skylion007 commented on torch/_dynamo/backends/train_step.py:
```diff
@@ -0,0 +1,197 @@
+
+            return [convert(idx, x) for idx, x in enumerate(flat_args)]
+
+        params = {
+            **dict(mod.named_parameters()),
+            **dict(mod.named_buffers()),
+        }
```
```suggestion
        params = dict(itertools.chain(
            mod.named_parameters(),
            mod.named_buffers()
        ))
```
will be more efficient or at least use dict.update

wconstab replied:
thanks- will add these in!

Skylion007 commented on torch/_dynamo/backends/train_step.py:
```diff
@@ -0,0 +1,197 @@
+                out = mod(*_user_args, **kwargs)
+
+            if not isinstance(out, (tuple, list)):
+                raise RuntimeError(
```
```suggestion
                raise TypeError(
```

wconstab committed:
Update on "Train Step Compile [1/3] basic fwd/bwd only"


Adds infra to wrap compilers in TrainStepCompiler and logic to dynamo
to behave specially (e.g. allowing .backward() to not graph-break)
when using such compilers.

Integrates some state with TracingContext, useful for both asserting
only one call to .backard() as well as groundwork for optimizer support
in following PR.

Checks various assumptions about the user trainstep function:
 - only one call to .backward() - double backwards is uncommon, and
   initial users of train step compile are expected to be satisfied
   without it.  Not sure if it would be feasible to support.
 - only supported backends are allowed
 - fullgraph (nopython) is required when train_step_compiling

Separated PRs for easier review, but this is of minimal utility
on its own without optimizer support.

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1946,6 +1947,9 @@ def __init__(
         # as soon as we create the tracing context we should keep it active, so any calls
         # into dynamo apis can rely on finding it
         with tracing(self.output.tracing_context):
+            if _is_train_step_compiler(compiler_fn):
```
Can we do this in a way that does not have `_is_train_step_compiler` in dynamo? this was the feedback in an earlier PR. https://github.com/pytorch/pytorch/pull/98461/

wconstab replied:
i don't know what you mean exactly.
1) you don't want any logic inside dynamo checking if we're in train step mode
2) its ok to have logic for train step mode, but you prefer it to take a different form (like a bool that comes from the user's api call)

I initially wrote it as (2) and hated how many lines of code i had to plumb "train_step=train_step" through to achieve this.  My current approach makes the change in just one place where it hands off from detecting the compile_fn type to marking the tracing context.

voznesenskym replied:
> you don't want any logic inside dynamo checking if we're in train step mode

Correct - dynamo should not know bout train step compiler

wconstab replied:
but... am I wrong to think that dynamo _has_ to know about train step compiler in order to decide to not graph-break on backwards?

voznesenskym commented on torch/_guards.py:
```diff
@@ -394,6 +394,19 @@ class TracingContext:
     will return None.
     """
 
+    class TrainStepContext:
```
Any context added needs to respect copy/restore.

wconstab replied:
could you help me write a comment that i could add to the TracingContext code that makes it more apparent what that means and how to do it?  Also maybe a test that confirms that TracingContexts are restorable?

wconstab committed:
Update on "Train Step Compile [1/3] basic fwd/bwd only"


Adds infra to wrap compilers in TrainStepCompiler and logic to dynamo
to behave specially (e.g. allowing .backward() to not graph-break)
when using such compilers.

Integrates some state with TracingContext, useful for both asserting
only one call to .backard() as well as groundwork for optimizer support
in following PR.

Checks various assumptions about the user trainstep function:
 - only one call to .backward() - double backwards is uncommon, and
   initial users of train step compile are expected to be satisfied
   without it.  Not sure if it would be feasible to support.
 - only supported backends are allowed
 - fullgraph (nopython) is required when train_step_compiling

Separated PRs for easier review, but this is of minimal utility
on its own without optimizer support.

[ghstack-poisoned]

wconstab committed:
Update on "Train Step Compile [1/3] basic fwd/bwd only"


Adds infra to wrap compilers in TrainStepCompiler and logic to dynamo
to behave specially (e.g. allowing .backward() to not graph-break)
when using such compilers.

Integrates some state with TracingContext, useful for both asserting
only one call to .backard() as well as groundwork for optimizer support
in following PR.

Checks various assumptions about the user trainstep function:
 - only one call to .backward() - double backwards is uncommon, and
   initial users of train step compile are expected to be satisfied
   without it.  Not sure if it would be feasible to support.
 - only supported backends are allowed
 - fullgraph (nopython) is required when train_step_compiling

Separated PRs for easier review, but this is of minimal utility
on its own without optimizer support.

[ghstack-poisoned]

wconstab commented on torch/_dynamo/backends/train_step.py:
```diff
@@ -0,0 +1,195 @@
+    train_step_fn: Callable,
+    *,
+    dynamic: builtins.bool = False,
+    backend: Union[str, Callable] = "inductor",
```
change default

ezyang commented:
Will and I had a 1:1 where we talked about the high level structure of these two PRs, and what the long term plan for them might be. We didn't manage to get through everything, but some major points:

* PR 2 modifies the passed in user optimizer *in-place* during compilation. This is bad because it violates a major invariant, which is that PT2 compile time **must not** modify userspace objects (we only ever modify userspace when we actually run our compiled outputs.) Potential fixes include copying the optimizer (but this is only OK for PyTorch library optimizers which we know our well-behaved in some way), or Dynamo bytecode tracing through the optimizer code (e.g., so when Dynamo process backward(), we need to know enough about backward() semantics to know how to populate VariableTracker fake tensors with appropriate grad fields.) Dynamo bytecode tracing through optimizer seems like a pretty powerful lever that it doesn't seem like we have explored enough; the current implementation is relying heavily on `make_fx` to trace through things instead, which is resulting in the problems we are seeing.
* Eager train-step-compile is a quite difficult API to deal with. Besides the problems we've already discussed in composability (optimizer initialization), it also has another problem of having a scope that is potentially too short: if you only control the train-step function, it is invalid to perform optimizations that move operations from the current train-step to the next train-step, because in eager mode the user can always change their mind and decide to not run the next train-step! Will agreed to check what distributed compiler is actually doing here. One way to solve this problem is to take over the entire Trainer, not just a single compile-step. Another option is to agree not to do cross-iteration optimizations.
* What does a potential long term design that doesn't require a separate train_step_compiler look like? Will had some intuition that higher order operators could help. The sketch of the design here is that a user is required to rewrite their train-step function to call a higher-order grad operator. This grad operator is maintained inside the Dynamo traced graph, and then AOTAutograd can trace through it the normal way (without requiring a special trace step that bypasses AOTAutograd). Will was not sure if this would work with optimizers and we didn't manage to fully address this.
* A big UX lever here is whether or not the train step compiler is export-style or eager-style. In an export-style regime, we don't offer to maintain the original user Module/Optimizer, instead everything is taken under the control of the export object. One sketch is the exported train-step function is a stateful blob that maintains parameters and optimizer states; you feed it inputs and it outputs losses and updates its internal optimizer state and parameters. In this regime, the black box would be responsible for initializing optimizer states (as it has taken control of them) and this is just one part of the exported object in question.
